### PR TITLE
Feature/updates

### PR DESCRIPTION
**Pull request by st3v3nhunt**
There is additional work for refining what gets mapped and ensuring there are no Out Of Memory exceptions occurring. However, there is a bunch of useful stuff available here which should be good to get merged in:
* the image having persistent data
* useful files be mounted via a volume
* doing all of the stuff within the image (as opposed to partly on the host)
* removing the generated file
* using some of the latest standards e.g. `/scripts`

The files in `./config` were taken straight from the base ES image. I've left them as there were which means there is a lot of useful(?) comments in there. Happy to discuss whether they should be stripped out.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,17 @@
 FROM elasticsearch:5.3.0
+RUN apt-get update && apt-get install -y \
```
Comment (neilbmclaughlin) [inline Dockerfile:2]: Odd place for a line break :)
Remove or put before the `&&` to give a command per line?

File: README.md [hunk 2]
```diff
@@ -1,9 +1,14 @@
-To create import document structure, removing opening times as this causes index problems run the following command:
+# Elasticsearch instance with GP profiles data
 
-jq -c '.[]  | .id = ._id | del (._id) | del (.openingTimes) | {"index": {"_index": "profiles", "_type": "gps", "_id": .id}}, .' data/gp-data-merged.json > data/gp-data-bulk-insert.json 
+Running [`scripts/start`](scripts/start) will format
```
Comment (neilbmclaughlin) [inline README.md]: This is no longer strictly true in that the work is done inside the container now.

File: data/scripts/load-data [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+until curl --output /dev/null --silent --head --fail http://127.0.0.1:9200; do
+    printf '.'
+    sleep 5
+done
+echo 'ES has started'
+
+echo 'Starting bulk data insert'
+curl -XPOST 127.0.0.1:9200/_bulk --data-binary @/usr/share/elasticsearch/tmp/data/output/gp-data-bulk-insert.json
+echo 'Finished bulk data insert'
+pkill -F /tmp/pid
```
Comment (neilbmclaughlin) [inline data/scripts/load-data]: Maybe a comment as to why the pkill and the wait above needs to be done?

File: data/scripts/load-data [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+echo 'Waiting for ES to start'
+until curl --output /dev/null --silent --head --fail http://127.0.0.1:9200; do
+    printf '.'
+    sleep 5
+done
+echo 'ES has started'
+
+echo 'Starting bulk data insert'
+curl -XPOST 127.0.0.1:9200/_bulk --data-binary @/usr/share/elasticsearch/tmp/data/output/gp-data-bulk-insert.json
```
Comment (neilbmclaughlin) [inline data/scripts/load-data]: The number lines of output from the bulk insert makes it unwieldy.

The response to the bulk insert is a json object which has a `took` and `errors` property as well as the `items` (see [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html)).

Could the response be piped through jq to return a summary (`took`, `errors` and `item.length`?)

File: scripts/bootstrap [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+echo "Set Docker host's max_map_count for ES"
+open "https://www.elastic.co/guide/en/elasticsearch/reference/current/vm-max-map-count.html"
```
Comment (neilbmclaughlin) [inline scripts/bootstrap]: Not sure what I need to do with the information on this page. Maybe something added to the readme about these settings?

File: Dockerfile [hunk 6]
```diff
@@ -1,10 +1,17 @@
...
 
-RUN chmod +x /tmp/load-data.sh && /tmp/load-data.sh
+COPY ./config/ /usr/share/elasticsearch/config/
+COPY ./data/ /usr/share/elasticsearch/tmp/data/
+
+RUN chown -R elasticsearch:elasticsearch /usr/share/elasticsearch/
 
 USER elasticsearch
+
+RUN /usr/share/elasticsearch/tmp/data/scripts/transform-data
```
Comment (neilbmclaughlin) [inline Dockerfile:16]: 💯  I think the idea of running the scripts which pre-processes the data for bulk insert within the container is a good thing.

**Review by neilbmclaughlin — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (st3v3nhunt) [reply Dockerfile:2]: Following best practices https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#run

File: data/scripts/load-data (hunk 3 above)
Comment (st3v3nhunt) [reply data/scripts/load-data]: Done

File: scripts/bootstrap (hunk 5 above)
Comment (st3v3nhunt) [reply scripts/bootstrap]: Added info into `README`

**Comment by st3v3nhunt** [comment]
@neilbmclaughlin feedback responded to with changes or explanatory comments

**Review by neilbmclaughlin — APPROVED** [review]

**Pull request merged**